Fix NullReferenceException if no IdentityServer key configured

martincostello opened the pull request:
**PR Title**

Fix NullReferenceException if no IdentityServer key configured

**PR Description**

Check that `SigningCredential` is not null when creating `InMemoryValidationKeysStore` to prevent a `NullReferenceException`.

Also fixes up the solution filter file and makes some small refactors I noticed while fixing the issue.

Addresses #32334

martincostello committed:
Add missing Identity projects

Add missing projects for running the Microsoft.AspNetCore.ApiAuthorization.IdentityServer.Tests project in Visual Studio.

martincostello committed:
Check SigningCredential is not null

Check that SigningCredential is not null when creating
InMemoryValidationKeysStore to prevent NRE.
Addresses #32334.

martincostello committed:
Make methods static

Make methods static to resolve Visual Studio suggestions.

martincostello committed:
Remove array allocation

Use overload that accepts a single char, rather than creating an array.

HaoK commented:
Looks reasonable, can you add a test for this?

martincostello commented:
> Looks reasonable, can you add a test for this?

Sure - I did look around at the existing test classes, but they were mostly low-level config validation rather than actually using the service collection extensions, so wasn't sure where to slot in a new one.

Shall I just add some unit tests for `IdentityServerBuilderConfigurationExtensions`? Validating the change will need building it into a service provider and then trying to resolve the service, so I felt it was starting to edge into an integration test more than a unit test.

HaoK commented:
Yeah a new class is fine, yeah just verifying that it does the right thing against a service collection which doesn't have the setting configured is the idea.

martincostello committed:
Add unit test for NRE fix

Add a unit test that validates the behaviour change in a6af0ed5e08bb131e757a75dfb3a277953bf063c.

martincostello committed:
Make IIdentityServerJwtDescriptor required

Make IIdentityServerJwtDescriptor a required dependency to prevent
a NullReferenceException being thrown if AddIdentityServerJwt() was
not added to the service configuration for the application.

martincostello commented:
Test added - in the process of adding that test, I also found that if you miss the call to `` in the service setup you also get a `NullReferenceException` from here due to the missing registration:

https://github.com/dotnet/aspnetcore/blob/8cf98856bb96c7aad8373ae5b1fa6e7f021f649e/src/Identity/ApiAuthorization.IdentityServer/src/Configuration/ConfigureApiResources.cs#L56

I've changed the code to make it throw an exception for the missing registration to try and give a hint to the problem, but alternatively I can change the method to guard against the dependency being null and revert the change to make the service required.

HaoK commented:
Ah nice, yeah I recall there were people running into that null ref when they forgot to call AddIdentityServerJwt before, so I think the better exception/require service is reasonable.  cc @javiercn in case he has a preference between the two options

javiercn commented:
> Ah nice, yeah I recall there were people running into that null ref when they forgot to call AddIdentityServerJwt before, so I think the better exception/require service is reasonable. cc @javiercn in case he has a preference between the two options

I think it's better to make it handle nulls values properly [here](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/ApiAuthorization.IdentityServer/src/Configuration/ConfigureApiResources.cs#L56) by just using `?.`

martincostello commented:
Ok cool, I'll change that.

martincostello committed:
Handle null rather than throw

Revert change from dbd3ea7a65c4efe3aaa5482b7aaa33af6ba03a6a and instead just handle it being null, rather than make it required.

HaoK commented on src/Identity/Identity.slnf:
```diff
@@ -86,7 +86,9 @@
       "src\\Http\\Metadata\\src\\Microsoft.AspNetCore.Metadata.csproj",
-      "src\\Localization\\Abstractions\\src\\Microsoft.Extensions.Localization.Abstractions.csproj"
+      "src\\Localization\\Abstractions\\src\\Microsoft.Extensions.Localization.Abstractions.csproj",
+      "src\\Components\\Web\\src\\Microsoft.AspNetCore.Components.Web.csproj",
```
Did you actually need the components project when working with identity here?  I think the other two additions are fine, but I'd rather not load components in the Identity slnf unless its really needed

martincostello replied:
Yes, it was needed to be referenced via the test project so that I could run the tests in Visual Studio, otherwise the test compilation failed.

HaoK reviewed (approved):
Looks good pending the one questions about the slnf file, thanks for doing this work @martincostello !

javiercn commented:
@HaoK merge it when you are happy
